src/hotspot/share/prims/jvmtiRedefineClasses.cpp

```diff
@@ -1379,22 +1378,21 @@ jvmtiError VM_RedefineClasses::load_new_class_versions() {
                        ClassFileStream::verify);
 
     // Parse the stream.
-    Handle the_class_loader(current, the_class->class_loader());
-    Handle protection_domain(current, the_class->protection_domain());
     // Set redefined class handle in JvmtiThreadState class.
     // This redefined class is sent to agent event handler for class file
     // load hook event.
     state->set_class_being_redefined(the_class, _class_load_kind);
 
     Thread* THREAD = current;  // for exception processing
     ExceptionMark em(THREAD);
+    Handle protection_domain(THREAD, the_class->protection_domain());
     ClassLoadInfo cl_info(protection_domain);
-    InstanceKlass* scratch_class = SystemDictionary::parse_stream(
-                                                      the_class_sym,
-                                                      the_class_loader,
-                                                      &st,
+    InstanceKlass* scratch_class = KlassFactory::create_from_stream(&st,
+                                                      the_class->name(),
+                                                      the_class->class_loader_data(),
                                                       cl_info,
                                                       THREAD);
+
```

Comment:
Could you add a comment above line 1390 saying you can't call resolve_class_from_stream() here because the resulting class should not go in the system dictionary?

Reply:
// Parse and create a class from the bytes, but this class isn't added
    // to the dictionary, so do not call resolve_from_stream.